Proposal: isDeprecated option for PropTypes

pavelkornev opened the pull request:
During lifetime some of the components props became deprecated. It would be nice to have a way to indicate these props in declarative way.

1. It has no conflicts with ```isRequired``` flag and can not be combined together since it makes no sense to require deprecated props.

2. It has full backward compatibility and will not break all existed components.

Examples of usage:
```
propTypes: {
  primitive: PropTypes.string.isDeprecated
  array: PropTypes.arrayOf(PropTypes.number).isDeprecated
  complex: PropTypes.oneOfType([PropTypes.string.isDeprecated, PropTypes.number]),
  shape: PropTypes.shape({
    key: PropTypes.oneOfType([PropTypes.string.isDeprecated, PropTypes.number])
  }).isDeprecated
}
```

As a result it will warn in console with the following message:
```
Deprecated prop `primitive` was specified in `MyComponent`.
```

satya164 commented:
I think this is not the ideal way to mark a prop as deprecated. When we mark a prop as deprecated, we should provide a clear message with instructions to how to migrate to the new API. In React Native we use a custom `deprecatedPropType` function to do this - https://github.com/facebook/react-native/blob/master/Libraries/Utilities/deprecatedPropType.js

Aweary commented:
Could `isDeprecated` be a function that takes an optional message describing how to upgrade?

```js
propTypes: {
  deprecated: PropTypes.string.isDeprecated(`
    'deprecated' is now deprecated, please use 'notDeprecated' instead.`
  ),
 notDeprecated: PropTypes.string
}
```

pavelkornev commented:
@satya164 @Aweary

1. I agree that deprecated message is essential and should be an explicit way to set it up. I've just added ability to pass message through.

2. I have realized that i've made a mistake in original commit and return an ```Error``` instance when ```isDeprecated``` was specified and didn't perform typeChecks. I've fixed it now and throw an ```console.warn``` which makes this function with side-effects and probably it's not the best way around. Some tests dropped down because of this change and will be fixed if we decide to merge this PR into master.

3. Function [deprecatedPropType](https://github.com/facebook/react-native/blob/master/Libraries/Utilities/deprecatedPropType.js) from React-Native mentioned by @satya164 looks nice. But is there analog in react itself we don't know about?

jimfb commented:
Thanks for the PR!

Proptypes are mostly legacy and in maintenance mode, having been replaced by tools like Flow, so we've generally shied away from adding new features/functions/api. There is no reason that this functionality needs to be tied into React, so my intuition is that the best strategy here is to just create a third-party library that helps users define deprecated proptypes, and post that on github.

Aweary commented:
You could just use a thin wrapper function that logs the warning and returns the `propType`

```js

function isDeprecated(propType, message) {
  console.warn(message)
  return propType
}

```

```js
propTypes: {
deprecated: isDeprecated(object, `
 'deprecated' is deprecated, please use 'nonDeprecated'
`)
}

```

Aweary commented:
I also just published [`react-is-deprecated`](https://github.com/Aweary/react-is-deprecated) which can add an `isDeprecated` method to your `React.PropTypes` object.

```js
import React from 'react';
import addIsDeprecated from 'react-is-deprecated';
const PropTypes = addIsDeprecated(React.PropTypes);

export default class SomeComponent extends React.Component {
  static propTypes = {
    deprecated: PropTypes.string.isDeprecated(`'deprecated' is deprecated, please use 'notDeprecated.'`),
    notDeprecated: PropTypes.string
  }
}
```

Seems to work OK, but feel free to fix it if it doesn't!

Aweary commented:
I've published [`react-is-deprecated`](https://github.com/Aweary/react-is-deprecated) which will can wrap `React.PropTypes` and provide an `isDeprecated` method.

```js
import React from 'react';
import addIsDeprecated from 'react-is-deprecated';
const PropTypes = addIsDeprecated(React.PropTypes);

export default class SomeComponent extends React.Component {
  static propTypes = {
    deprecated: PropTypes.string.isDeprecated(`'deprecated' is deprecated, please use 'notDeprecated.'`),
    notDeprecated: PropTypes.string
  }
}
```

gaearon commented:
Thanks to everyone in the discussion.

I’d say this will better live in user land, at least initially.
As @Aweary wrote above, he made a package for it: https://github.com/Aweary/react-is-deprecated.

If the package proves to be popular enough we can reconsider adding this to the core.
For now I’ll close this PR but I’m very grateful to @pavelkornev for starting the discussion.

Aweary commented:
Thanks @gaearon for the mention and the feedback and @pavelkornev for the inspiration. Feel free to open an issue/PR if there's anything you think we can do to make it a better package :+1:
